File: packages/backend/src/jobs/queue.ts
import { and, asc, eq, inArray, lte, sql } from 'drizzle-orm';
import { Cron } from 'croner';
import { db } from '../database';
import { jobs } from '../database/schema';
import { config } from '../config';
import { logger } from '../config/logger';

/**
 * In-process job queue over the app's SQLite database (successor to
 * BullMQ→pg-boss). Single process + synchronous better-sqlite3 means claiming
 * is inherently race-free; crash recovery is a boot-time reset of 'active'
 * rows. Processors and their dispatch-by-name contract are unchanged.
 */

export type QueueName = 'ingestion' | 'indexing' | 'remote-content';
export const QUEUE_NAMES: QueueName[] = ['ingestion', 'indexing', 'remote-content'];

/** What processors receive — the (BullMQ-era) Job surface they actually use. */
export interface QueueJob<T = any> {
	id: string;
	name: string;
	data: T;
}

export interface SendOptions {
	/** Retries after the first attempt. 0 = run exactly once. Default 4 (= 5 tries). */
	retryLimit?: number;
	/** Suppress this send while a job with the same key is pending or running. */
	singletonKey?: string;
	/** Do not run before this time. */
	startAfter?: Date;
}

// Master/dispatcher jobs (initial-import, continuous-sync) create a sync session
// and fan out per-mailbox jobs; they are NOT idempotent, so a retry would create
// a duplicate session and re-dispatch every mailbox. Run them exactly once —
// unlike the idempotent batch jobs (index-email-batch, archive-remote-content-batch)
// which keep the default retry policy for transient failures.
export const masterJobOptions: SendOptions = { retryLimit: 0 };

type Dispatcher = (job: QueueJob) => Promise<unknown>;

interface QueueRuntime {
	dispatcher: Dispatcher;
	concurrency: number;
	running: number;
}

const runtimes = new Map<QueueName, QueueRuntime>();
const inFlight = new Set<Promise<void>>();
let poller: NodeJS.Timeout | null = null;
let retentionSweeper: NodeJS.Timeout | null = null;
let cron: Cron | null = null;
let stopping = false;

const POLL_MS = 500;
const BACKOFF_BASE_MS = 1000;
const BACKOFF_CAP_MS = 5 * 60 * 1000;

/** Registers the consumer for a queue (called by workers/index.ts at boot). */
export const registerWorker = (
	queue: QueueName,
	dispatcher: Dispatcher,
	concurrency: number
): void => {
	runtimes.set(queue, { dispatcher, concurrency, running: 0 });
};

const claimAndRun = (queue: QueueName, runtime: QueueRuntime): void => {
	const free = runtime.concurrency - runtime.running;
	if (free <= 0) {
		return;
	}
	const now = new Date();
	const claimable = db
		.select()
		.from(jobs)
		.where(and(eq(jobs.queue, queue), eq(jobs.state, 'pending'), lte(jobs.runAt, now)))
		.orderBy(asc(jobs.createdAt))
		.limit(free)
		.all();
	for (const job of claimable) {
		db.update(jobs)
			.set({ state: 'active', startedAt: now, attempts: job.attempts + 1 })
			.where(eq(jobs.id, job.id))
			.run();
		runtime.running++;
		const run = runtime
			.dispatcher({ id: job.id, name: job.name, data: job.payload })
			.then(() => {
				db.update(jobs)
					.set({ state: 'completed', finishedAt: new Date(), error: null })
					.where(eq(jobs.id, job.id))
					.run();
			})
			.catch((error: unknown) => {
				const attempts = job.attempts + 1;
				const message =
					error instanceof Error ? (error.stack ?? error.message) : String(error);
				if (attempts >= job.maxAttempts) {
					db.update(jobs)
						.set({ state: 'failed', finishedAt: new Date(), error: message })
						.where(eq(jobs.id, job.id))
						.run();
					logger.error({ queue, name: job.name, jobId: job.id, error: message }, 'Job failed permanently');
				} else {
					const delay = Math.min(BACKOFF_BASE_MS * 2 ** (attempts - 1), BACKOFF_CAP_MS);
					db.update(jobs)
						.set({ state: 'pending', runAt: new Date(Date.now() + delay), error: message })
						.where(eq(jobs.id, job.id))
						.run();
					logger.warn({ queue, name: job.name, jobId: job.id, attempts, delay }, 'Job failed, retrying');
				}
			})
			.finally(() => {
				runtime.running--;
				inFlight.delete(run);
			});
		inFlight.add(run);
	}
};

const tick = (): void => {
	if (stopping) {
		return;
	}
	for (const [queue, runtime] of runtimes) {
		try {
			claimAndRun(queue, runtime);
		} catch (error) {
			logger.error({ queue, error }, 'Queue tick failed');
		}
	}
};

/** Boot-time start: recover crash leftovers and begin polling. */
export const startQueue = async (): Promise<void> => {
	stopping = false;
	// Anything 'active' at boot is a leftover from a crash/kill — re-run it.
	const recovered = db
		.update(jobs)
		.set({ state: 'pending', runAt: new Date() })
		.where(eq(jobs.state, 'active'))
		.run();
	if (recovered.changes > 0) {
		logger.warn({ count: recovered.changes }, 'Recovered interrupted jobs from previous run');
	}
	poller = setInterval(tick, POLL_MS);
	// Retention: completed jobs age out after 2 days, failed after 14.
	retentionSweeper = setInterval(
		() => {
			try {
				db.run(sql`
					DELETE FROM jobs
					WHERE (state = 'completed' AND finished_at < (unixepoch() * 1000) - ${2 * 24 * 3600 * 1000})
						OR (state = 'failed' AND finished_at < (unixepoch() * 1000) - ${14 * 24 * 3600 * 1000})
				`);
			} catch (error) {
				logger.error({ error }, 'Job retention sweep failed');
			}
		},
		10 * 60 * 1000
	);
	retentionSweeper.unref();
	logger.info('Job queue started (SQLite, in-process)');
};

/** Graceful stop: cease claiming, wait for in-flight jobs (bounded by caller). */
export const stopQueue = async (): Promise<void> => {
	stopping = true;
	cron?.stop();
	cron = null;
	if (poller) {
		clearInterval(poller);
		poller = null;
	}
	if (retentionSweeper) {
		clearInterval(retentionSweeper);
		retentionSweeper = null;
	}
	await Promise.allSettled([...inFlight]);
};

/**
 * Enqueues a named job. Returns the job id, or null when suppressed by
 * options.singletonKey (a job with the same key is already pending/active).
 */
export const sendJob = async (
	queue: QueueName,
	name: string,
	payload: object,
	options: SendOptions = {}
): Promise<string | null> => {
	if (options.singletonKey) {
		const existing = db
			.select({ id: jobs.id })
			.from(jobs)
			.where(
				and(
					eq(jobs.queue, queue),
					eq(jobs.singletonKey, options.singletonKey),
					inArray(jobs.state, ['pending', 'active'])
				)
			)
			.limit(1)
			.all();
		if (existing.length > 0) {
			logger.info(
				{ queue, name, singletonKey: options.singletonKey },
				'Duplicate job suppressed (already queued or running).'
			);
			return null;
		}
	}
	const [row] = db
		.insert(jobs)
		.values({
			queue,
			name,
			payload,
			maxAttempts: (options.retryLimit ?? 4) + 1,
			runAt: options.startAfter ?? new Date(),
			singletonKey: options.singletonKey ?? null,
		})
		.returning({ id: jobs.id })
		.all();
	// Nudge the poller so locally-enqueued work starts without the poll delay.
	setImmediate(tick);
	return row?.id ?? null;
};

/**
 * Registers the continuous-sync cron (in-process via croner; pattern comes from
 * SYNC_FREQUENCY). The singletonKey makes a tick that fires while the previous
 * one is still queued a no-op.
 */
export const registerSyncSchedule = async (): Promise<void> => {
	cron?.stop();
	cron = new Cron(config.app.syncFrequency, () => {
		void sendJob(
			'ingestion',
			'schedule-continuous-sync',
			{},
			{ retryLimit: 0, singletonKey: 'schedule-continuous-sync' }
		);
	});
	logger.info({ pattern: config.app.syncFrequency }, 'Continuous sync schedule registered');
};

/**
 * Removes queued/retryable/failed ingestion jobs belonging to a source (force
 * sync cleanup). Active jobs are left alone — same net behavior as before.
 */
export const removeJobsBySourceId = async (ingestionSourceId: string): Promise<number> => {
	const result = db.run(sql`
		DELETE FROM jobs
		WHERE queue = 'ingestion'
			AND state IN ('pending', 'failed')
			AND json_extract(payload, '$.ingestionSourceId') = ${ingestionSourceId}
	`);
	const count = Number(result.changes ?? 0);
	logger.info({ ingestionSourceId, count }, 'Removed queued jobs for source during force sync.');
	return count;
};

/** Builds a queue dispatcher from a named-job processor map. */
export const makeDispatcher = (
	processors: Record<string, (job: QueueJob) => Promise<unknown>>
): Dispatcher => {
	return async (job: QueueJob): Promise<void> => {
		const processor = processors[job.name];
		if (!processor) {
			throw new Error(`Unknown job name: ${job.name}`);
		}
		await processor(job);
	};
};

File: crates/engine/src/queue.rs
//! Port of jobs/queue.ts — the in-process SQLite job queue: claim loop with
//! per-queue concurrency, exponential backoff (1s·2^n, cap 5min), singleton
//! suppression, boot recovery of interrupted jobs, and the retention sweep.

use crate::state::AppState;
use rusqlite::Connection;
use serde_json::Value;
use std::collections::HashMap;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;

pub const QUEUE_NAMES: [&str; 3] = ["ingestion", "indexing", "remote-content"];
const POLL_MS: u64 = 500;
const BACKOFF_BASE_MS: i64 = 1000;
const BACKOFF_CAP_MS: i64 = 5 * 60 * 1000;

fn now_ms() -> i64 {
    crate::search::now_ms()
}

pub struct SendOptions<'a> {
    /// Retries after the first attempt. 0 = run exactly once. Default 4.
    pub retry_limit: i64,
    pub singleton_key: Option<&'a str>,
    pub start_after: Option<i64>,
}

impl Default for SendOptions<'_> {
    fn default() -> Self {
        Self { retry_limit: 4, singleton_key: None, start_after: None }
    }
}

pub fn master_job_options(singleton_key: &str) -> SendOptions<'_> {
    SendOptions { retry_limit: 0, singleton_key: Some(singleton_key), start_after: None }
}

/// A one-shot job: run exactly once. File imports are single-mailbox and
/// deterministic, so a failed process-mailbox won't recover on the queue's
/// backoff — and retrying is actively broken, because the first failure is
/// `is_last` and finalizes (deletes) the import session, so every retry then
/// errors with "session not found". One shot → clean terminal 'error' state.
pub fn no_retry() -> SendOptions<'static> {
    SendOptions { retry_limit: 0, singleton_key: None, start_after: None }
}

/// sendJob — returns the job id, or None when singleton-suppressed.
pub fn send_job(
    state: &AppState,
    queue: &str,
    name: &str,
    payload: &Value,
    options: SendOptions,
) -> Option<String> {
    let conn = state.pool.get().unwrap();
    if let Some(key) = options.singleton_key {
        let existing: Option<String> = conn
            .query_row(
                "SELECT id FROM jobs WHERE queue = ? AND singleton_key = ? \
                 AND state IN ('pending', 'active') LIMIT 1",
                rusqlite::params![queue, key],
                |r| r.get(0),
            )
            .ok();
        if existing.is_some() {
            return None;
        }
    }
    let id = uuid::Uuid::new_v4().to_string();
    if let Err(e) = conn.execute(
        "INSERT INTO jobs (id, queue, name, payload, max_attempts, run_at, singleton_key) \
         VALUES (?, ?, ?, ?, ?, ?, ?)",
        rusqlite::params![
            id,
            queue,
            name,
            payload.to_string(),
            options.retry_limit + 1,
            options.start_after.unwrap_or_else(now_ms),
            options.singleton_key,
        ],
    ) {
        // Enqueue failed (e.g. transient SQLITE_BUSY). Log rather than panic:
        // send_job runs inside worker threads, and a panic there would unwind
        // past the concurrency-slot release and permanently stall the queue.
        // `None` means "not enqueued" here (callers ignore the return value).
        eprintln!("[queue] failed to enqueue {queue}/{name}: {e}");
        return None;
    }
    state.queue_nudge.notify_one();
    Some(id)
}

/// removeJobsBySourceId — clears queued/failed ingestion jobs for a source.
pub fn remove_jobs_by_source_id(conn: &Connection, source_id: &str) -> i64 {
    conn.execute(
        "DELETE FROM jobs WHERE queue = 'ingestion' AND state IN ('pending', 'failed') \
         AND json_extract(payload, '$.ingestionSourceId') = ?",
        [source_id],
    )
    .map(|n| n as i64)
    .unwrap_or(0)
}

struct ClaimedJob {
    id: String,
    queue: String,
    name: String,
    payload: Value,
    attempts: i64,
    max_attempts: i64,
}

fn claim(conn: &Connection, queue: &str, free: i64) -> Vec<ClaimedJob> {
    if free <= 0 {
        return Vec::new();
    }
    let now = now_ms();
    let mut stmt = conn
        .prepare(
            "SELECT id, name, payload, attempts, max_attempts FROM jobs \
             WHERE queue = ? AND state = 'pending' AND run_at <= ? \
             ORDER BY created_at ASC LIMIT ?",
        )
        .unwrap();
    let candidates: Vec<ClaimedJob> = stmt
        .query_map(rusqlite::params![queue, now, free], |row| {
            Ok(ClaimedJob {
                id: row.get(0)?,
                queue: queue.to_string(),
                name: row.get(1)?,
                payload: serde_json::from_str(&row.get::<_, String>(2)?)
                    .unwrap_or(Value::Null),
                attempts: row.get(3)?,
                max_attempts: row.get(4)?,
            })
        })
        .unwrap()
        .filter_map(Result::ok)
        .collect();
    // Flip pending→active atomically and only return jobs we actually won. The
    // `AND state = 'pending'` guard + rows-affected check means a job whose claim
    // write is lost (or was already taken) is left for a later tick, never run
    // while still 'pending' — otherwise the next poll could claim and run it again.
    let mut claimed = Vec::new();
    for job in candidates {
        match conn.execute(
            "UPDATE jobs SET state = 'active', started_at = ?, attempts = ? \
             WHERE id = ? AND state = 'pending'",
            rusqlite::params![now, job.attempts + 1, job.id],
        ) {
            Ok(1) => claimed.push(job),
            _ => {}
        }
    }
    claimed
}

fn finish(conn: &Connection, job: &ClaimedJob, result: Result<(), String>) {
    // Retry the state write briefly. A silently dropped completion write would
    // leave the job 'active' and re-run it after the next restart; busy_timeout
    // already waits out most contention, so a couple of retries closes the gap.
    let durable = |sql: &str, params: &[&dyn rusqlite::ToSql]| {
        for attempt in 0..3 {
            if conn.execute(sql, params).is_ok() {
                return;
            }
            if attempt < 2 {
                std::thread::sleep(std::time::Duration::from_millis(100));
            }
        }
        eprintln!("[queue] {}:{} state write failed after retries", job.queue, job.name);
    };
    match result {
        Ok(()) => durable(
            "UPDATE jobs SET state = 'completed', finished_at = ?, error = NULL WHERE id = ?",
            rusqlite::params![now_ms(), job.id],
        ),
        Err(message) => {
            let attempts = job.attempts + 1;
            if attempts >= job.max_attempts {
                durable(
                    "UPDATE jobs SET state = 'failed', finished_at = ?, error = ? WHERE id = ?",
                    rusqlite::params![now_ms(), message, job.id],
                );
                eprintln!("[queue] {}:{} failed permanently: {message}", job.queue, job.name);
            } else {
                let delay = (BACKOFF_BASE_MS << (attempts - 1).min(20)).min(BACKOFF_CAP_MS);
                durable(
                    "UPDATE jobs SET state = 'pending', run_at = ?, error = ? WHERE id = ?",
                    rusqlite::params![now_ms() + delay, message, job.id],
                );
            }
        }
    }
}

/// Boot-time recovery: an 'active' job is a leftover from a crash. Re-queue it
/// only if it still has attempts left; a job that had already used its last
/// attempt (e.g. a no_retry process-mailbox that crashed after committing) is
/// marked 'failed' rather than re-run, which would double-count its work.
pub fn recover_interrupted(conn: &Connection) {
    conn.execute(
        "UPDATE jobs SET \
           state = CASE WHEN attempts >= max_attempts THEN 'failed' ELSE 'pending' END, \
           run_at = ?, \
           finished_at = CASE WHEN attempts >= max_attempts THEN ? ELSE finished_at END \
         WHERE state = 'active'",
        rusqlite::params![now_ms(), now_ms()],
    )
    .ok();
}

fn retention_sweep(conn: &Connection) {
    conn.execute_batch(&format!(
        "DELETE FROM jobs WHERE \
         (state = 'completed' AND finished_at < (unixepoch() * 1000) - {}) \
         OR (state = 'failed' AND finished_at < (unixepoch() * 1000) - {})",
        2 * 24 * 3600 * 1000i64,
        14 * 24 * 3600 * 1000i64
    ))
    .ok();
}

fn concurrency_for(queue: &str) -> i64 {
    match queue {
        "ingestion" => std::env::var("INGESTION_WORKER_CONCURRENCY")
            .ok()
            .and_then(|v| v.parse().ok())
            .unwrap_or(5),
        "indexing" => 1,
        "remote-content" => 2,
        _ => 1,
    }
}

/// The worker loop — poll + nudge driven; each claimed job runs on the
/// blocking pool through the processor dispatch in `crate::processors`.
pub async fn start_queue(state: AppState) {
    if let Ok(conn) = state.pool.get() {
        recover_interrupted(&conn);
    }
    // One-shot cleanup of stale tracker "failures" recorded before the tracking
    // filter recognized them (runs off the async executor; self-limiting).
    {
        let state = state.clone();
        tokio::task::spawn_blocking(move || crate::remote_content::sweep_tracking_assets(&state));
    }
    let running: Arc<HashMap<&'static str, AtomicI64>> = Arc::new(
        QUEUE_NAMES.iter().map(|q| (*q, AtomicI64::new(0))).collect(),
    );

    // Retention + stale-session sweep every 10 minutes. (The Node engine's
    // continuous-sync cron is gone: every provider is a one-time file import,
    // so the minutely schedule only produced no-op or unprocessable jobs.)
    {
        let state = state.clone();
        tokio::spawn(async move {
            let mut interval = tokio::time::interval(std::time::Duration::from_secs(600));
            loop {
                interval.tick().await;
                if let Ok(conn) = state.pool.get() {
                    retention_sweep(&conn);
                    crate::sessions::clean_stale_sessions(&conn);
                }
            }
        });
    }

    let mut interval = tokio::time::interval(std::time::Duration::from_millis(POLL_MS));
    loop {
        tokio::select! {
            _ = interval.tick() => {},
            _ = state.queue_nudge.notified() => {},
        }
        for queue in QUEUE_NAMES {
            let counter = &running[queue];
            let free = concurrency_for(queue) - counter.load(Ordering::SeqCst);
            if free <= 0 {
                continue; // fully busy — also avoids a `LIMIT <= 0` claim
            }
            // A transient pool-checkout timeout must NOT kill this loop: it is
            // spawned unsupervised, so an `.unwrap()` panic here would permanently
            // stop all background processing. Skip this tick and retry next poll.
            let jobs = match state.pool.get() {
                Ok(conn) => claim(&conn, queue, free),
                Err(_) => continue,
            };
            for job in jobs {
                counter.fetch_add(1, Ordering::SeqCst);
                let state = state.clone();
                let running = running.clone();
                tokio::task::spawn_blocking(move || {
                    // Catch a panicking job so it is marked failed (not left
                    // 'active') AND the concurrency slot is ALWAYS released — a
                    // panic escaping here would leak the slot and stall the queue.
                    let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                        crate::processors::dispatch(&state, &job.queue, &job.name, &job.payload)
                    }))
                    .unwrap_or_else(|_| Err(format!("job {}/{} panicked", job.queue, job.name)));
                    // Record the outcome if we can, but ALWAYS release the slot and
                    // nudge afterwards — a failed pool checkout here must not leak
                    // the slot (which would stall this queue for the whole run). A
                    // job left 'active' by a missed write is re-queued on next boot.
                    match state.pool.get() {
                        Ok(conn) => finish(&conn, &job, result),
                        Err(e) => eprintln!(
                            "[queue] could not record result for {}/{}: {e}",
                            job.queue, job.name
                        ),
                    }
                    running[job.queue.as_str()].fetch_sub(1, Ordering::SeqCst);
                    state.queue_nudge.notify_one();
                });
            }
        }
    }
}

/// Synchronous drain for the CLI import: claim and run jobs single-threaded
/// until every queue is empty of runnable work. Backoff-delayed retries are
/// honoured by sleeping to the nearest run_at.
pub fn drain_for_cli(state: &AppState) -> Result<(), String> {
    loop {
        let mut ran_any = false;
        for queue in QUEUE_NAMES {
            loop {
                let jobs = {
                    let conn = state.pool.get().map_err(|e| e.to_string())?;
                    claim(&conn, queue, 1)
                };
                let Some(job) = jobs.into_iter().next() else { break };
                let result = crate::processors::dispatch(state, &job.queue, &job.name, &job.payload);
                let conn = state.pool.get().map_err(|e| e.to_string())?;
                finish(&conn, &job, result);
                ran_any = true;
            }
        }
        if ran_any {
            continue;
        }
        // Nothing claimable now — check for future-scheduled pending work.
        let next_run: Option<i64> = {
            let conn = state.pool.get().map_err(|e| e.to_string())?;
            conn.query_row(
                "SELECT min(run_at) FROM jobs WHERE state = 'pending'",
                [],
                |r| r.get(0),
            )
            .ok()
            .flatten()
        };
        match next_run {
            Some(at) => {
                let wait = (at - now_ms()).max(0) as u64;
                std::thread::sleep(std::time::Duration::from_millis(wait.min(5000) + 10));
            }
            None => return Ok(()),
        }
    }
}
